tests/scripts/generate_test_code.py

```diff
@@ -0,0 +1,837 @@
+        if len(s) == 2:
+            code = """exit:
+    ;;
+}""".join(s)
+
+    code += gen_function_wrapper(name, locals, args_dispatch)
+    ifdef, endif = gen_deps(deps)
+    dispatch_code = gen_dispatch(name, suite_deps + deps)
+    return name, args, ifdef + code + endif, dispatch_code
+
+
+def parse_functions(funcs_f):
+    """
+    Parses a test_suite_xxx.function file and returns information
+    for generating a C source file for the test suite.
+
+    :param funcs_f: file object of the functions file.
+    :return: List of test suite dependencies, test function dispatch
+             code, function code and a dict with function identifiers
+             and arguments info.
+    """
+    suite_headers = ''
+    suite_helpers = ''
+    suite_deps = []
+    suite_functions = ''
+    func_info = {}
+    function_idx = 0
+    dispatch_code = ''
+    for line in funcs_f:
+        if re.search(BEGIN_HEADER_REGEX, line):
+            headers = parse_until_pattern(funcs_f, END_HEADER_REGEX)
+            suite_headers += headers
+        elif re.search(BEGIN_SUITE_HELPERS_REGEX, line):
+            helpers = parse_until_pattern(funcs_f, END_SUITE_HELPERS_REGEX)
+            suite_helpers += helpers
+        elif re.search(BEGIN_DEP_REGEX, line):
+            deps = parse_suite_deps(funcs_f)
+            suite_deps += deps
+        elif re.search(BEGIN_CASE_REGEX, line):
+            deps = parse_function_deps(line)
+            func_name, args, func_code, func_dispatch =\
+                parse_function_code(funcs_f, deps, suite_deps)
+            suite_functions += func_code
+            # Generate dispatch code and enumeration info
+            if func_name in func_info:
+                raise GeneratorInputError(
+                    "file: %s - function %s re-declared at line %d" % \
+                    (funcs_f.name, func_name, funcs_f.line_no))
+            func_info[func_name] = (function_idx, args)
+            dispatch_code += '/* Function Id: %d */\n' % function_idx
+            dispatch_code += func_dispatch
+            function_idx += 1
+
+    ifdef, endif = gen_deps(suite_deps)
+    func_code = ifdef + suite_headers + suite_helpers + suite_functions + endif
+    return suite_deps, dispatch_code, func_code, func_info
+
+
+def escaped_split(str, ch):
+    """
+    Split str on character ch but ignore escaped \{ch}
+    Since, return value is used to write back to the intermediate
+    data file, any escape characters in the input are retained in the
+    output.
+
+    :param str: String to split
+    :param ch: split character
+    :return: List of splits
+    """
+    if len(ch) > 1:
+        raise ValueError('Expected split character. Found string!')
+    out = []
+    part = ''
+    escape = False
+    for i in range(len(str)):
+        if not escape and str[i] == ch:
+            out.append(part)
+            part = ''
+        else:
+            part += str[i]
+            escape = not escape and str[i] == '\\'
```

Comment:
There was some good reason to not use `re.split`. I think it was that escaping the '\' didn't work with it:
```
>>> re.split(r'[^\\]:', 'alkaa:assa')
['alka', 'assa']
>>> re.split(r'[^\\]:', 'alkaa\:assa')
['alkaa\\:assa']
>>> re.split(r'[^\\]:', 'alkaa\\:assa')
['alkaa\\:assa']
```

Reply:
Actually it is a complex scenario than it looks. It is hard to make a regex and specify escape the escape character.

Reply:
Indeed I don't think you can use `re.split` when the separator can be escaped. However you can take advantage of the fact that the string can't contain `\n`, so you can translate the separator to `\n`. As long as you document what you're doing, I think it makes the code a lot more readable.

    # Replace the field separator ':' by a newline, which can't be present in the string.
    # Replace backslash-character sequences by the quoted character.
    # Then split on newlines (which were originally unquoted separators).
    out = sub(r'\\.|:', lambda s: '\n' if s == ':' else s[1], str).split('\n')

Reply:
I meant

    out = re.sub(r'\\(.)|:', lambda m: m.group(1) or '\n', str, len(str)).split('\n')

Oh, and please don't use `str` as a variable name. It's the name of a built-in module.

Reply:
Ah nice. I didn't pay attention to `repl`, that can be a callable. Since, I need to retain the escape sequence for writing to the .datax file. I will group the whole `\\.`:
```
>>> re.sub(r'(\\.)|:', lambda m: m.group(1) or '\n', r"ABC:DEF\:PQR", len(r'ABC:DEF\:PQR')).split('\n')
['ABC', 'DEF\\:PQR']
```